### DISCUSSION TITLE
contributing/development/compiling/introduction_to_the_buildsystem

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#215] [User-contributed notes]
# contributing/development/compiling/introduction_to_the_buildsystem

Godot is a primarily C++ project and it uses the SCons build system. We love SCons for how maintainable and easy to set up it makes our buildsystem. And thanks to that compiling Godot from source c...

https://docs.godotengine.org/en/latest/contributing/development/compiling/introduction_to_the_buildsystem.html

**Comment by ghost** [comment]
`git clean` can be used with `--exclude=<file/folder>` to preserve some files. It is useful if you want to make clean build but don't want to rewrite your `custom.py` file so you can just run `git clean -fixd --exclude=custom.py`

**Comment by enetheru** [comment]
This line from the **Target** heading was pointed out to me today:

> `target=template_debug`: Build with C++ debugging symbols (defines: `DEBUG_ENABLED`, `-O2`/`/O2`)

I don't think its correct, from reading SConscript file, the debug symbol flags are predicated on the `debug_symbols` flag. `dev_build` also implies `debug_symbols` unless overridden. But I see no path from `template_debug` to `debug_symbols`.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10647

**Comment by jinyangcruise** [comment]
When I define a local variable with the name of a member variable, scons shows some messy codes in my computer:
```
...
editor\find_in_files.cpp(1177): error C2220: ���¾��汻��Ϊ����
editor\find_in_files.cpp(1177): warning C4458: ��tabs_visible�����������������Ա
.\scene/gui/tab_container.h(49): note: �μ���TabContainer::tabs_visible��������
scons: *** [bin\obj\editor\find_in_files.windows.editor.x86_64.obj] Error 2
scons: building terminated because of errors.
```
I know why error occurs but I don't know why the information became messy code. Can anyone help me?
I use `scons platform=windows dev_mode=yes` to build.
My computer is Windows11 and the system language is Chinese.

**Reply by Calinou** [reply]
Unlike GCC/MinGW, Visual Studio reports compiler errors in your system language. I assume this output is going through SCons for processing and it's not handling the non-ASCII characters well. (I don't know if there's an easy way to force it to report compiler errors in English.)

Can you open an issue on <https://github.com/godotengine/godot/issues>?

**Reply by jinyangcruise** [reply]
Thx, I opened an issue https://github.com/godotengine/godot/issues/104710

**Comment by mvuets** [comment]
> ## Using multi-threading
> [...] By default, Godot's SCons setup is configured to use all CPU threads but one (to keep the system responsive during compilation).

This isn't the case on my system. I got 4 CPU threads (over 2 cores) and all 4 are utilised by default:

```
$ scons
scons: Reading SConscript files ...
Automatically detected platform: linuxbsd
Auto-detected 4 CPU cores available for build parallelism. Using 4 cores by default. You can override it with the `-j` or `num_jobs` arguments.
```

I believe [N - 1 threads are used only when total is > 4](https://github.com/godotengine/godot/blob/9b22b41531c82132732fda0f9818c5709c2587bc/SConstruct#L576).

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11081
